File: src/foam/swift/parse/json/MapParser.js
/**
 * @license
 * Copyright 2017 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */

foam.CLASS({
  package: 'foam.swift.parse.json',
  name: 'MapParser',
  extends: 'foam.swift.parse.parser.ProxyParser',
  requires: [
    'foam.swift.parse.json.AnyKeyParser',
    'foam.swift.parse.json.Whitespace',
    'foam.swift.parse.parser.Literal',
    'foam.swift.parse.parser.Repeat',
    'foam.swift.parse.parser.Seq0',
    'foam.swift.parse.parser.Seq1',
    'foam.swift.parse.parser.Seq2',
  ],
  axioms: [
    foam.pattern.Singleton.create()
  ],
  properties: [
    {
      name: 'delegate',
      swiftFactory: function() {/*
return 
  Seq1_create(["index": 2, "parsers": [
    Whitespace_create(),
    Literal_create(["string": "{"]),
    Repeat_create([
      "delegate":
        Seq2_create(["index1": 1, "index2": 5, "parsers": [
          Whitespace_create(),
          AnyKeyParser_create(),
          Whitespace_create(),
          Literal_create(["string": ":"]),
          Whitespace_create(),
          __context__.create(foam_swift_parse_json_AnyParser.self)!,
        ]]),
      "delim":
        Seq0_create(["parsers": [
          Whitespace_create(),
          Literal_create(["string": ","]),
        ]])
    ]),
    Whitespace_create(),
    Literal_create(["string": "}"]),
  ]])
      */},
    },
  ],
  methods: [
    {
      name: 'parse',
      swiftCode: function() {/*
let ps = super.parse(ps, x)
if ps != nil {
  let values = ps!.value() as! [Any?]
  var map: [String:Any?] = [:]
  for item in values {
    let item = item as! [Any?]
    map[item[0] as! String] = item[1]
  }
  return ps!.setValue(map)
}
return ps
      */},
    },
  ],
});

File: src/foam/lib/json/MapParser.java
/**
 * @license
 * Copyright 2017 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */

package foam.lib.json;

import foam.lib.parse.*;

public class MapParser
  extends ObjectNullParser
{
  private final static Parser instance__ = new MapParser();

  public static Parser instance() { return instance__; }
  private MapParser() {
    super(new Seq1(3,
      Whitespace.instance(),
      Literal.create("{"),
      Whitespace.instance(),
      new Repeat(new Seq2(1, 5,
        Whitespace.instance(),
        new AnyKeyParser(),
        Whitespace.instance(),
        Literal.create(":"),
        Whitespace.instance(),
        AnyParser.instance()),
        new Seq0(Whitespace.instance(), Literal.create(","))),
      Whitespace.instance(),
      Literal.create("}")));
  }

  public PStream parse(PStream ps, ParserContext x) {
    ps = super.parse(ps, x);
    if ( ps == null ) return ps;

    Object[] values = (Object[]) ps.value();
    java.util.Map map = new java.util.HashMap(values.length);

    for ( int i = 0 ; i < values.length ; i++ ) {
      Object[] item = (Object[]) values[i];

      map.put(item[0], item[1]);
    }

    return ps.setValue(map);
  }
}
